header.c

```diff
@@ -0,0 +1,2084 @@
+                h_tag->len = j-i;
+                h_tag->next = NULL;
+                if (!h_tag->str)
+                    return -1;
+
+                if (last)
+                    last->next = h_tag;
+                else
+                    h_type->tag = h_tag;
+
+                last = h_tag;
+                i = j;
+            } while (i < len && hdr[i] != '\0' && hdr[i] != '\n');
+        }
+
+        /* Update RG/SQ hashes */
+        if (-1 == bam_hrecs_update_hashes(hrecs, type, h_type))
+            return -1;
+    }
+
+    return 0;
+}
+
+/*! Update bam_hdr_t target_name and target_len arrays
+ *
+ *  @return 0 on success; -1 on failure
+ */
+int update_target_arrays(bam_hdr_t *bh, const bam_hrecs_t *hrecs,
+                         int refs_changed) {
+    if (!bh || !hrecs)
+        return -1;
+
+    if (refs_changed < 0)
+        return 0;
+
+    // Grow arrays if necessary
+    if (bh->n_targets < hrecs->nref) {
+        char **new_names = realloc(bh->target_name,
+                                   hrecs->nref * sizeof(*new_names));
+        if (!new_names)
+            return -1;
+        bh->target_name = new_names;
+        uint32_t *new_lens = realloc(bh->target_len,
+                                     hrecs->nref * sizeof(*new_lens));
+        if (!new_lens)
+            return -1;
+        bh->target_len = new_lens;
+    }
+
+    // Update names and lengths where changed
+    // hrecs->refs_changed is the first ref that has been updated, so ones
+    // before that can be skipped.
+    int i;
+    for (i = refs_changed; i < hrecs->nref; i++) {
+        if (i >= bh->n_targets
+            || strcmp(bh->target_name[i], hrecs->ref[i].name) != 0) {
+            if (i < bh->n_targets)
+                free(bh->target_name[i]);
+            bh->target_name[i] = strdup(hrecs->ref[i].name);
+            if (!bh->target_name[i])
+                return -1;
+        }
+        bh->target_len[i] = hrecs->ref[i].len;
+    }
+
+    // Free up any names that have been removed
+    for (; i < bh->n_targets; i++) {
+        free(bh->target_name[i]);
+    }
+
+    bh->n_targets = hrecs->nref;
+    return 0;
+}
+
+static int rebuild_target_arrays(bam_hdr_t *bh) {
+    if (!bh || !bh->hrecs)
+        return -1;
+
+    bam_hrecs_t *hrecs = bh->hrecs;
+    if (hrecs->refs_changed < 0)
+        return 0;
+
+    if (update_target_arrays(bh, hrecs, hrecs->refs_changed) != 0)
+        return -1;
+
+    hrecs->refs_changed = -1;
+    return 0;
+}
+
+int bam_hdr_parse(bam_hdr_t *bh) {
+    bam_hrecs_t *hrecs = bam_hrecs_new();
+
+    if (!hrecs)
+        return -1;
+
+    // Parse existing header text
+    if (bh->text && bh->l_text > 0) {
+        if (bam_hrecs_parse_lines(hrecs, bh->text, bh->l_text) != 0) {
+            bam_hrecs_free(hrecs);
+            return -1;
+        }
+    }
+
+    bh->hrecs = hrecs;
+
+    if (hrecs->refs_changed >= 0 && rebuild_target_arrays(bh) != 0)
+        return -1;
+
+    bam_hdr_link_pg(bh);
```

Comment:
Calling `bam_hdr_link_pg` would result in resetting `bh->text`, which is not intuitive when calling `bam_hdr_parse`, or some of its callers that do not modify the header, such as `bam_hdr_name2ref`.
For example, the following code would result in a null text field, which is probably not the expected behaviour.
```
        samFile *in = sam_open("test_header.sam", "r");
        bam_hdr_t *h = sam_hdr_read(in);	
        int id = bam_hdr_name2ref(h, "ref2");
        printf("id=%d, header=%p\n", id, h->text);

Perhaps, a better place for the `bam_hdr_link_pg` call would be inside `bam_hdr_rebuild`.